docs: Add IAVL v1 migration guide for consensus node operators (#4915)
celestiaorg/celestia-app: docs/maintainers/iavl-v1-migration.md

> # IAVL v1 Migration Guide for Consensus Nodes
> 
> This guide helps consensus node operators migrate from IAVL v0 to IAVL v1 to achieve optimal performance benefits.

Reviewer: ```suggestion
# IAVL v1 Migration Guide for Consensus Nodes
```

too much wording here, the title is self explanatory

## Background

> Starting with celestia-app v4, the state machine uses IAVL v1.x instead of v0.x. IAVL v1 introduces data locality optimizations that provide roughly **10x performance improvement** over the previous version.

Reviewer: ```suggestion
Starting with celestia-app v4, the state machine can use IAVL v1.x instead of v0.x. IAVL v1.
```

### Key Benefits of IAVL v1

- **Significantly faster state access**: ~10x performance improvement
- **Better data locality**: More efficient disk I/O patterns

> - **Reduced state access bottlenecks**: Prevents state access from limiting throughput

Reviewer: redo this and don't harp on the benefits that much, we only care about informing node operators what they need to do to switch. focus on that.

### Migration Considerations

The entire database needs to be migrated to the v1 key layout. There are two migration approaches:

1. **Lazy Migration** (default): Automatic migration during normal operation
2. **State Sync Migration** (recommended): Full migration via state sync for optimal performance

## Prerequisites

Before starting the migration process:

1. **Verify you're ready for v4**: Ensure you understand the v4 upgrade requirements from the [release notes](../release-notes/release-notes.md)
2. **Backup critical data**: Always backup validator keys and state before any migration
3. **Plan downtime**: Especially for state sync migration, plan for node downtime
4. **Monitor resources**: Ensure sufficient disk space and I/O capacity for migration
5. **Test environment**: If possible, test the migration process on a non-production node first

## Migration Options

### Option 1: Lazy Migration (Default)

When you upgrade to celestia-app v4, IAVL v1 migration happens automatically and lazily as your node operates. 

> **Pros:**
> - Requires no additional action from operators
> - Node continues operating normally during migration
> 
> **Cons:**
> - May cause increased I/O load during the migration period
> - State access might become a performance bottleneck initially
> - Not all data is immediately optimized for v1 layout
> 
> **When to use:**
> - You want a simple upgrade process
> - Your node can tolerate temporary performance degradation
> - You're not concerned about maximizing throughput immediately

Reviewer: don't claim things you don't know
```suggestion
**Pros:**
- Node continues operating normally during migration

**Cons:**
- May cause increased I/O load during the migration period

**When to use:**
- you want a simple upgrade process
```

### Option 2: State Sync Migration (Recommended)

Perform a fresh state sync with celestia-app v4 to ensure all state data uses the new IAVL v1 key layout from the start.

> **Pros:**
> - Immediate access to full IAVL v1 performance benefits
> - All state data is optimized for the new layout
> - No performance degradation period
> 
> **Cons:**
> - Requires more manual intervention
> - Longer initial setup time
> - Requires backing up and restoring validator keys

Reviewer: ```suggestion
**Pros:**
- Immediate access to full IAVL v1 performance benefits

**Cons:**
- Requires more manual intervention
```

> **When to use:**
> - You want maximum performance immediately
> - Your node requires optimal throughput
> - You can afford the setup time for state sync

Reviewer: ```suggestion
**When to use:**
- You want maximum performance immediately
```

   ```

2. **Stop your current node**:
   ```bash
   # Stop the celestia-app service
   sudo systemctl stop celestia-appd
   ```

3. **Upgrade to celestia-app v4**:
   ```bash
   # If installed from source (with multiplexer support)
   git checkout v4.x.x  # Replace with actual v4 release tag
   make install
   
   # For standalone binary without multiplexer (if already on v4 network)
   # make install-standalone
   
   # If using prebuilt binaries, download from releases page
   # wget https://github.com/celestiaorg/celestia-app/releases/download/v4.x.x/celestia-app_Linux_x86_64.tar.gz
   ```

   > **Note**: celestia-app v4 includes a multiplexer by default that supports syncing from genesis using embedded v3 binaries. The multiplexer will automatically switch to v4 state machine when the network upgrades. If your network has already upgraded to v4, you can use the standalone binary.

4. **Start your node with v4**:
   ```bash
   # Start the service
   sudo systemctl start celestia-appd
   
   # Monitor logs for migration progress
   journalctl -u celestia-appd -f
   ```

5. **Monitor performance**: Watch for completion of lazy migration and performance improvements over time.

> ### State Sync Migration Process

Reviewer: remove this section, as operators already know how to state sync

   ```

4. **Configure state sync** in `~/.celestia-app/config/config.toml`:
   ```toml
   [statesync]
   enable = true
   
   # Use trusted RPC endpoints for state sync
   # For mainnet:
   rpc_servers = "https://rpc-1.celestia.org:443,https://rpc-2.celestia.org:443"
   
   # For testnet, use appropriate testnet RPC endpoints
   # rpc_servers = "https://rpc-mocha.pops.one:443,https://rpc.celestia-mocha.com:443"
   
   # Get trust height and hash from a recent block
   # You can query: curl -s https://rpc-1.celestia.org/block | jq -r '.result.block.header.height + "," + .result.block_id.hash'
   trust_height = 0  # Replace with actual height
   trust_hash = ""   # Replace with actual hash
   
   # Optional: Configure snapshot settings
   trust_period = "112h"  # 2/3 of unbonding period
   ```

5. **Start celestia-app v4**:
   ```bash
   # Start the service
   sudo systemctl start celestia-appd
   
   # Monitor state sync progress
   journalctl -u celestia-appd -f
   ```

6. **Verify state sync completion**:
   ```bash
   # Check if the node is catching up
   celestia-appd status | jq .SyncInfo.catching_up
   
   # Should return false when sync is complete
   ```

> ## Configuration Options

Reviewer: remove this section

celestia-app v4 provides several IAVL-related configuration options in `app.toml` that can help optimize performance:

```toml
# IavlCacheSize set the size of the iavl tree cache. 
# Default cache size is 50mb.
iavl-cache-size = 781250

# IavlDisableFastNode enables or disables the fast node feature of IAVL. 
# Default is false.
iavl-disable-fastnode = false

# EXPERIMENTAL: IAVLLazyLoading enable/disable the lazy loading of iavl store.
# Default is false.
iavl-lazy-loading = false
```

### Configuration Recommendations

- **iavl-cache-size**: Increase this value if you have sufficient RAM. A larger cache can improve performance for frequently accessed state.
- **iavl-disable-fastnode**: Keep this as `false` (default) to benefit from IAVL v1's fast node features.
- **iavl-lazy-loading**: This is experimental. Test carefully before enabling in production.

## Performance Monitoring

> ### Before Migration

Reviewer: remove these sections as well
